external/llvm-project/mlir/lib/Conversion/AMDGPUToROCDL/AMDGPUToROCDL.cpp

```diff
@@ -305,49 +306,100 @@ static Value mfmaConcatIfNeeded(ConversionPatternRewriter &rewriter,
   return input;
 }
 
-/// Return the `rocdl` intrinsic corresponding to a `MFMAInstr` value.
-/// This conversion happens here to allow code up the stack to handle the choice
-/// of mfma by picking between enum variants, which is much more ergonomic than
-/// picking between ops, at the cost of some long switch statements in this
-/// pass.
-static StringRef mfmaInstrToIntrinsicName(MFMAInstr instr) {
-#define LOWERING_CASE(type)                                                    \
-  case MFMAInstr::type:                                                        \
-    return ROCDL::mfma_##type::getOperationName();
-  switch (instr) {
-    LOWERING_CASE(f32_32x32x1f32)
-    LOWERING_CASE(f32_16x16x1f32)
-    LOWERING_CASE(f32_4x4x1f32)
-    LOWERING_CASE(f32_32x32x2f32)
-    LOWERING_CASE(f32_16x16x4f32)
-    LOWERING_CASE(f32_32x32x4f16)
-    LOWERING_CASE(f32_16x16x4f16)
-    LOWERING_CASE(f32_4x4x4f16)
-    LOWERING_CASE(f32_32x32x8f16)
-    LOWERING_CASE(f32_16x16x16f16)
-    LOWERING_CASE(i32_32x32x4i8)
-    LOWERING_CASE(i32_16x16x4i8)
-    LOWERING_CASE(i32_4x4x4i8)
-    LOWERING_CASE(i32_32x32x8i8)
-    LOWERING_CASE(i32_16x16x16i8)
-    LOWERING_CASE(f32_32x32x2bf16)
-    LOWERING_CASE(f32_16x16x2bf16)
-    LOWERING_CASE(f32_4x4x2bf16)
-    LOWERING_CASE(f32_32x32x4bf16)
-    LOWERING_CASE(f32_16x16x8bf16)
-    LOWERING_CASE(f32_32x32x4bf16_1k)
-    LOWERING_CASE(f32_16x16x4bf16_1k)
-    LOWERING_CASE(f32_4x4x4bf16_1k)
-    LOWERING_CASE(f32_32x32x8bf16_1k)
-    LOWERING_CASE(f32_16x16x16bf16_1k)
-    LOWERING_CASE(f64_16x16x4f64)
-    LOWERING_CASE(f64_4x4x4f64)
-    LOWERING_CASE(i32_16x16x32_i8)
-    LOWERING_CASE(i32_32x32x16_i8)
-    LOWERING_CASE(f32_16x16x8_xf32)
-    LOWERING_CASE(f32_32x32x4_xf32)
+/// Return the `rocdl` intrinsic corresponding to a MFMA operation `mfma`
+/// if one exists. This includes checking to ensure the intrinsic is supported
+/// on the architecture you are compiling for.
+static Optional<StringRef> mfmaOpToIntrinsic(MFMAOp mfma, Chipset chipset) {
+  uint32_t m = mfma.m(), n = mfma.n(), k = mfma.k(), b = mfma.nBlks();
+  Type sourceElem = mfma.sourceA().getType();
+  if (auto sourceType = sourceElem.dyn_cast<VectorType>())
+    sourceElem = sourceType.getElementType();
+  Type destElem = mfma.destC().getType();
+  if (auto destType = destElem.dyn_cast<VectorType>())
+    destElem = destType.getElementType();
+
```

Comment:
Could we rename `nBlks` to `outBlks`, and instead of use `nBlks` as input, use it as a derived parameter from `bool reduction`. I think this is going to impact a number of things from your PR.

Derive m from n or n from m, make sure they are always equivalent.

Such that we can compute the nBlks using: 
 - If nonreduction, this must satisfy: `nBlks x m(or n) == 64`
   - If chipset >= 0x40, `nBlks x m(or n) == 128` 
 - if reduction: `nBlks = 1`

For example, in fp32:
 - 32x32: nBlocks = 2
 - 16x16x1 (nonreduction): nBlks = 4
 - 16x16x4 (reduction): nBlks = 1

Reply:
I'm not sure that if statement is accurate - many of the gfx940 ops still have reduction-less ops that have (m/n) * nBlks = 64

Reply:
And there are reduction cases where `nBlks` is not 1, IIRC. Also, the number of input and output blocks is always equal, so `nBlks` (or `blocks`, or `numBlocks`) is better than `outputBlocks`

Reply:
Ah, concrete example of k > 1, nBlks > 1, `mfma_f32_32x32x4f16`

Reply:
> I'm not sure that if statement is accurate - many of the gfx940 ops still have reduction-less ops that have (m/n) * nBlks = 64

That's right, I guess for backward compatibility purpose it has to make a shorter argument xdlops available.

> Ah, concrete example of k > 1, nBlks > 1, mfma_f32_32x32x4f16

For this case m = 32, nBlks = 2, so it satisfies 32 x 2 = 64.

> And there are reduction cases where nBlks is not 1, IIRC. Also, the number of input and output blocks is always equal.

For existing code base, input blocks and output blocks means different things and can be unequal (xdlops selection). For example, 
 - mfma_f32_16x16x4f16 is considered to have 4 input blocks and 4 output blocks
 - mfma_f32_16x16x1f16 is considered to have 4 input blocks and 1 output block, that's what made it an reduction versus the first one.

So when you say input and output blocks are always equal which part of code base are you referring to?

Do you see a straightforward way to incorporate `reduction` as a parameter at this abstraction layer?

Reply:
I'm referring to the individual `mfma` instructions themselves. They have parameters `m`, `n`, `k`, `blocks` (which is probably a better name than `nBlks`) and types. 

So, the point of this whole setup is that an individual MFMA is not selected by name but instead by m/n/k/blocks, which makes upstream happier.

I could absolutely see `bool reduction;` being a basic object at the level of xdlops selection, but I don't think it'll work correctly as part of `amdgpu.mfma`.

Reply:
Gotcha, so blocks is effectively `num_input_blks` in the old context.

> but I don't think it'll work correctly as part of amdgpu.mfma

No I think you are good. I was debating yesterday whether blocks is the more "fundamental" information but now I see each can be derived from the other parameter. Plus reduction does sound like a derived information on the abstraction level higher than mfma instructions.